### PR TITLE
setting pipeline for project

### LINKED ISSUE #14: Create CI for the project

### PR DESCRIPTION
**Pull request by ndefokou**
# 🚀 CI/CD Infrastructure & Codebase Stabilization for AgentFlow

## 📌 Summary
This PR introduces a robust CI/CD infrastructure for the **AgentFlow workspace** and stabilizes the codebase by resolving all existing linting warnings.

It establishes strict quality gates to ensure long-term **maintainability**, **security**, and **stability** across all crates.

---

### 1. ⚙️ CI/CD Infrastructure

- **Core CI Pipeline (`ci.yml`)**
  - Automates:
    - `cargo fmt`
    - `cargo clippy`
    - build & test workflows
  - Enforces a **Zero-Warning policy** across the workspace

- **Dependency Security (`cargo_deny.yml` & `deny.toml`)**
  - Automated checks for:
    - Security vulnerabilities
    - License compliance
    - Crate bans

- **API Stability (`semver_checks.yml`)**
  - Prevents accidental breaking changes
  - Ensures semantic versioning compliance

- **Standard Tooling**
  - `cargo-machete` → detects unused dependencies  
  - `typos` → spellchecking across the codebase  

---

### 2. 🧹 Stability & Refactors

- **Workspace-wide Cleanup**
  - Fixed numerous Clippy warnings:
    - Redundant imports
    - Dead code
    - Inefficient string operations
    - Nested formatting issues

- **Persistent File Locking**
  - Refactored `FileLockManager` in `pair-harness`
  - Maintains persistent file handles
  - Fixes race conditions and flaky tests in isolation logic

- **Handoff Logic Improvements**
  - Enhanced markdown parser in the `reset` module
  - Handles multiple metadata styles and edge cases during state transitions

- **Demo Improvements**
  - Improved build reliability
  - Reduced unnecessary output noise

---

### 3. 🧪 Testing & Verification

- Resolved flaky tests across:
  - Isolation
  - `watchdog`
  - `reset`

- Updated E2E test assertions:
  - Aligned with structured output formats

**Result:**
- 30 / 30 tests passing  
- 0 Clippy warnings (strict mode)

---

### 🔹 Local Build

```bash
cargo clippy --workspace --all-targets --all-features -- -D warnings
cargo test --workspace --all-features

### CODE REVIEW

File: deny.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    "MIT",
+    "Apache-2.0",
+    "BSD-2-Clause",
+    "BSD-3-Clause",
+    "ISC",
+    "CC0-1.0",
+    "Unicode-3.0",
+    "CDLA-Permissive-2.0",
+    "Zlib",
+    "Apache-2.0 WITH LLVM-exception",
```
Comment (Christiantyemele) [inline deny.toml]: we have just one license for now, please remove these ones

**Comment by Christiantyemele** [comment]
## PR #29 Review: CI/CD Infrastructure & Codebase Stabilization

### Overall: REQUEST_CHANGES — solid work with a few issues to address

---

### Positive Aspects

- **CI pipeline is well-designed** — `ci.yml` covers format, clippy, build, test (nextest), docs, unused deps, and typos. Good choice using nextest.
- **Security audit** — `cargo_deny.yml` with daily cron + `deny.toml` is best practice.
- **Semver checks** — Proactive API stability enforcement is excellent.
- **FileLockManager fix** (`isolation.rs`) — Most impactful bug fix. Holding `File` handles in `active_locks` correctly prevents premature `flock` release. Real correctness fix.
- **Unused dependency cleanup** — 12 unused deps removed across crates.
- **Idiomatic Rust improvements** — `#[default]` on `TicketStatus::Open`, eta-reduction, `push('\n')` vs `push_str("\n")`, `map` vs `filter_map`.

---

### Must Fix

1. **`semver_checks.yml` hardcodes package list** — Silently skips new crates (`agent-sentinel`, `anthropic-mock`). Remove the `package` filter to check all workspace crates, or add a comment noting the list must be maintained.

2. **`deny.toml` ignores `RUSTSEC-2026-0097`** without a tracking plan. Ignored advisories tend to accumulate. Add a TODO or tracking issue.

3. **`thiserror = "1"` in workspace `Cargo.toml`** — Only `thiserror 2.0.18` is used now (v1 removed from Cargo.lock). Update workspace dep to `"2"` or remove if unused.

4. **`ci.yml` missing caching** — No `Swatinem/rust-cache` or equivalent. CI will be slow for this many crates.

5. **`crate-ci/typos@master`** — Floating tag is not reproducible. Pin to a specific release (e.g., `v1.30.0`).

---

### Should Fix

6. **Commit history** — All 8 commits have `"setting pipeline for project"` or `"mergin main"`. Makes bisecting difficult. Use descriptive messages going forward.

7. **`watchdog.rs` sleep 1100ms → 2100ms** — Band-aid fix. Consider `tokio::sync::Barrier` or notification channel instead of relying on longer sleeps.

8. **`agent-nexus` removed `github` dep** — Verify NexusNode doesn't rely on GitHub operations. No explanation in PR description.

---

### Nitpicks

9. PR title is vague — `"ci: add CI/CD pipeline and resolve clippy warnings"` would be better.
10. `deny.toml` missing `[sources]` section — best practice to restrict allowed registries.
11. `cargo_deny.yml` `--all-features` may fail if features are mutually exclusive.

---

After addressing items 1–5, this is ready to merge.

File: binary/src/bin/demo.rs [hunk 2]
```diff
@@ -91,27 +93,7 @@ async fn main() -> Result<()> {
...
-        let _action = if current_node == "nexus" {
-            // We need to access the node directly, but Flow abstracts it.
-            // For the demo, let's just use the flow.run but with tracing on.
-            break; // We'll just use flow.run and let the background info logs show it
-        } else {
-            break;
-        };
-    }
-
-    // Actually, flow.run already has good tracing. Let's just use it and add more info level logs.
```
Comment (Christiantyemele) [inline binary/src/bin/demo.rs:114]: is this change breaking the flow ?, can please add it back for safety

**Review by Christiantyemele — CHANGES_REQUESTED** [review]

**Review by Christiantyemele — APPROVED** [review]
